### PR TITLE
Notebooks: Partial Fix for Bracket Closing Issue in Strings

### PR DESCRIPTION
**Pull request by chlafreniere**
#7080 occurs in a rare case.

Take this JSON:
 ```
           "outputs": [
                {
                    "name": "stdout",
                    "text": "\u001b[0m",
                    "output_type": "stream"
                }
            ],
```

In the NotebookTextFileModel, we leverage the TextEditorModel.matchBracket() API. For the .ipynb file type, the notebook extension references a copy of the JSON language-configuration.json:

```
	"brackets": [
		["{", "}"],
		["[", "]"]
	],
	"autoClosingPairs": [
		{ "open": "{", "close": "}", "notIn": ["string"] },
		{ "open": "[", "close": "]", "notIn": ["string"] },
		{ "open": "(", "close": ")", "notIn": ["string"] },
		{ "open": "'", "close": "'", "notIn": ["string"] },
		{ "open": "/*", "close": "*/", "notIn": ["string"] },
		{ "open": "\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "`", "close": "`", "notIn": ["string", "comment"] }
	]
```

However, this API doesn't ignore brackets that are don't have pairs in a string literal.

So, the bug here is that, in this particular notebook, we try to find the closing bracket (given the opening bracket location for the "outputs" property in JSON), and we end up finding the closing bracket for something not corresponding to outputs at all:

`...u001b[0...",`

(This ends up matching with the last ']' in the file)

Note: this isn't the full fix for this issue; it's possible to have a string that has multiple opening (or closing) brackets. One solution may be to dynamically keep track of the location for a closing bracket, but that's a risky change, especially for an issue that is fairly rare.

While this bug is bad, I'm not convinced that it's a mainline scenario. If you save the notebook (instead of using "Save As", this bug doesn't repro at all.

### CODE REVIEW

**Comment by corivera** [comment]
Does the **"notIn": ["string"]** part of the json configuration not cover this?

**Comment by chlafreniere** [comment]
One would think. But in the debugger it matches them incorrectly

**Comment by corivera** [comment]
Could this possibly be an upstream problem with vscode, then?

File: src/sql/workbench/parts/notebook/common/models/notebookTextFileModel.ts [hunk 1]
```diff
@@ -191,7 +193,9 @@ export class NotebookTextFileModel {
 			}
 		}
 		let outputsEnd = textEditorModel.textEditorModel.matchBracket({ column: outputsBegin.endColumn - 1, lineNumber: outputsBegin.endLineNumber });
-		if (!outputsEnd || outputsEnd.length < 2) {
+		// The end of any cell's outputs cannot possibly be the second to last line in the text editor model
+		// The second to last line (']') is always the end of the "cells" array. This is a temporary solution to a larger bracket matching issue.
+		if (!outputsEnd || outputsEnd.length < 2 || outputsEnd[1].endLineNumber === textEditorModel.textEditorModel.getLineCount() - 1) {
 			return undefined;
```
Comment (corivera) [inline src/sql/workbench/parts/notebook/common/models/notebookTextFileModel.ts]: What does the saved notebook end up looking like after this change? Do we not include outputs with brackets?

**Comment by chlafreniere** [comment]
I'll have a comprehensive fix out soon.

**Pull request closed without merging**